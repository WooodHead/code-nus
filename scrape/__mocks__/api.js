const api = jest.genMockFromModule('../api');
const DEFAULT_RESPONSES = require('./fixtures/responses.json');

let graphqlResponses = [];
function setGraphqlResponses(responses = DEFAULT_RESPONSES) {
  graphqlResponses = responses.slice();
}

const graphqlApi = {
  // eslint-disable-next-line
  request(query, variables) {
    return new Promise((resolve, reject) => {
      process.nextTick(
        () =>
          graphqlResponses.length
            ? resolve(graphqlResponses.shift())
            : reject(new Error('No response found')),
      );
    });
  },
};

api.graphqlApi = graphqlApi;
api.setGraphqlResponses = setGraphqlResponses;

module.exports = api;
