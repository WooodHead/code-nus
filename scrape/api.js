require('dotenv').config();

const GitHubApi = require('github');
const { GraphQLClient } = require('graphql-request');

const token = process.env.GITHUB_TOKEN;
if (!token) {
  throw new Error('No token found. Make sure .env file exists!');
}

const restApi = new GitHubApi({
  debug: process.env.NODE_ENV !== 'production',
});

restApi.authenticate({
  type: 'token',
  token,
});

const graphqlApi = new GraphQLClient('https://api.github.com/graphql', {
  headers: {
    Authorization: `bearer ${process.env.GITHUB_TOKEN}`,
  },
});

module.exports = {
  restApi,
  graphqlApi,
};
