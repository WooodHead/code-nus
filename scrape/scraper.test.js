jest.mock('write-file-atomic');
jest.mock('./api');

const writeFileAtomic = require('write-file-atomic');
const Scraper = require('./scraper');

const scraper = new Scraper();

describe('scraper', () => {
  beforeEach(() => {
    // eslint-disable-next-line
    require('./api').setGraphqlResponses();
  });

  it('should scrape from root repos for forks', async () => {
    const results = await scraper.getAllForks();
    expect(results).toHaveLength(3);
  });

  it('should process scaped data cleanly and remove duplicates', async () => {
    const results = await scraper.getAllForks();
    const processed = Scraper.processForks(results);
    expect(processed).toHaveLength(2);
    // stringify then parse to get rid of undefined values
    expect(JSON.parse(JSON.stringify(processed, null, 2))).toMatchSnapshot();
  });

  it('should write to file accordingly', async () => {
    await scraper.scrapeAsJson();
    expect(writeFileAtomic).toHaveBeenCalled();
  });
});
