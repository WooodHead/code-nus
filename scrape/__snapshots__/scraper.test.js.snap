// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scraper should process scaped data cleanly and remove duplicates 1`] = `
Array [
  Object {
    "avatarUrl": "https://avatars3.githubusercontent.com/u/7464833?v=4",
    "login": "sdfsdfsdf",
    "name": "Chan Jun Wei",
    "repositories": Array [
      Object {
        "descriptionHTML": "<div> Creating multiple angular cli app to modularize the project</div>",
        "homepageUrl": "",
        "nameWithOwner": "chanjunweimy/angular-multiple-app",
        "primaryLanguage": "TypeScript",
        "repositoryTopics": Array [],
        "stargazers": 7,
        "url": "https://github.com/chanjunweimy/angular-multiple-app",
      },
      Object {
        "descriptionHTML": "<div>Demo of using multiple angular applications and aspnetboilerplate to create plugin application</div>",
        "homepageUrl": "",
        "nameWithOwner": "chanjunweimy/abp_plugin_with_ui",
        "primaryLanguage": "C#",
        "repositoryTopics": Array [
          "aspnetboilerplate",
          "angular4",
          "angular-cli",
        ],
        "stargazers": 4,
        "url": "https://github.com/chanjunweimy/abp_plugin_with_ui",
      },
      Object {
        "descriptionHTML": "<div></div>",
        "homepageUrl": null,
        "nameWithOwner": "chanjunweimy/CBARS",
        "primaryLanguage": "Java",
        "repositoryTopics": Array [],
        "stargazers": 3,
        "url": "https://github.com/chanjunweimy/CBARS",
      },
      Object {
        "descriptionHTML": "<div>Search Machine</div>",
        "homepageUrl": null,
        "nameWithOwner": "chanjunweimy/SearchMachine",
        "primaryLanguage": "Java",
        "repositoryTopics": Array [],
        "stargazers": 1,
        "url": "https://github.com/chanjunweimy/SearchMachine",
      },
    ],
  },
  Object {
    "avatarUrl": "https://avatars3.githubusercontent.com/u/8993333?v=4",
    "login": "YukiXueli",
    "name": null,
    "repositories": Array [
      Object {
        "descriptionHTML": "<div></div>",
        "homepageUrl": null,
        "nameWithOwner": "YukiXueli/image-cache-and-sync",
        "primaryLanguage": "Java",
        "repositoryTopics": Array [],
        "stargazers": 0,
        "url": "https://github.com/YukiXueli/image-cache-and-sync",
      },
    ],
  },
]
`;
